### PR TITLE
Merge

### PR DESCRIPTION
**Pull request by sithmel**
Feedbacks are welcome

### CODE REVIEW

File: src/index.d.ts [hunk 1]
```diff
@@ -243,6 +243,9 @@ export declare function iterable<T> (iterator: { next: () => {value: T} } | Iter
 export declare function map<T, O> (func: (item: T) => O): (iter: Iterable<T>) => IterableIterator<O>
 export declare function map<T, O> (func: (item: T) => O, iter: Iterable<T>): IterableIterator<O>
 
+export declare function merge<T> (pickFunc: (items: Array<{ done: boolean, value: T } | null>) => number): (iterables: Array<Iterable<T>>) => IterableIterator<T>
```
Comment (KSXGitHub) [inline src/index.d.ts]: `iterables` should be `ReadonlyArray` instead of `Array`.

Comment (KSXGitHub) [reply src/index.d.ts]: Also, `items` should also be discriminated union of readonly objects:

```typescript
items: ReadonlyArray<
  { readonly done: false, readonly value: T } |
  { readonly done: true } |
  null
>
```

File: src/__test__/merge.test.js [hunk 2]
```diff
@@ -0,0 +1,82 @@
+/* eslint-env node, jest */
+const { merge, asyncMerge, asyncToArray, enumerate, asyncThrottle } = require('iter-tools')
+
+describe.only('merge', function () {
```
Comment (sithmel) [inline src/__test__/merge.test.js]: oh my. I have to fix that

**Comment by sithmel** [comment]
@KSXGitHub I think I am done here. Can you please have a look at TS ?
I am sure I forgot something, I am trying to get better at it, but it will take a bit, given that I am not using it in my day to day work.

**Comment by conartist6** [comment]
I think the thing to do (experimenting with it in sequins) is to just make the tests Typescript files. That way you know the common usages have types that work.

**Comment by sithmel** [comment]
I have tried this:
```ts
import assert from 'static-type-assert'
import * as iter from '../index'

assert<
  IterableIterator<1 | 2 | 3 | 4>
>(iter.merge(iter.mergeByComparison((a, b) => a - b), [[1, 2], [3, 4]]))
```
But I don't honestly understand the error message ...
```
src/__spec__/merge.spec.ts:6:3 - error TS2345: Argument of type 'IterableIterator<number>' is not assignable to parameter of type 'IterableIterator<2 | 4 | 3 | 1>'.
  Type 'number' is not assignable to type '2 | 4 | 3 | 1'.

6 >(iter.merge(iter.mergeByComparison((a, b) => a - b), [[1, 2], [3, 4]]))
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by KSXGitHub** [comment]
@sithmel By default, TypeScript interprets `[0, 1, 2]` as `number[]` instead of literal `[0, 1, 2]`. Try this:

```diff
 import assert from 'static-type-assert'
 import * as iter from '../index'

 assert<
   IterableIterator<1 | 2 | 3 | 4>
->(iter.merge(iter.mergeByComparison((a, b) => a - b), [[1, 2], [3, 4]]))
+>(iter.merge(iter.mergeByComparison((a, b) => a - b), [[1, 2], [3, 4]] as [[1, 2], [3, 4]]))
```

**Comment by sithmel** [comment]
I managed to have a minimal type test running. I am not sure it is any good, though

**Pull request merged**